File: bot/services/anticheat.py
"""Жауаптардың сенімділігін тексеру — anti-cheat жүйесі."""

import statistics


def validate_answers(answers: list, timings: list) -> dict:
    """Жауаптарды тексеру және anti-cheat нәтижелерін қайтару."""
    result = {}

    option_indices = [a.get("option_index", 0) for a in answers]

    # 1. Straight-line detection — бірдей жауаптар қатарынан
    result["straight_line_count"] = _longest_consecutive_run(option_indices)

    # 2. Speed checks — тым жылдам жауаптар
    fast_threshold_ms = 2000
    result["fast_answers"] = sum(1 for t in timings if t < fast_threshold_ms)
    result["avg_time_ms"] = statistics.mean(timings) if timings else 0

    # 3. Low variance — тегтердің әртүрлілігі
    from collections import Counter
    all_tags = []
    for a in answers:
        all_tags.extend(a.get("tags", []))
    tag_counts = Counter(all_tags)
    if len(tag_counts) > 1:
        values = list(tag_counts.values())
        result["variance_score"] = statistics.stdev(values) / statistics.mean(values)
    else:
        result["variance_score"] = 0.0

    # 4. Pattern cycling — қайталанатын паттерн
    result["pattern_detected"] = _detect_cycling_pattern(option_indices)

    return result


def calculate_confidence(anticheat: dict) -> int:
    """Anti-cheat нәтижелерінен сенімділік пайызын есептеу (0-100)."""
    confidence = 100

    # Straight-line: ≥4 қатарынан бірдей → штраф
    if anticheat["straight_line_count"] >= 4:
        confidence -= (anticheat["straight_line_count"] - 3) * 15

    # Жылдам жауаптар: әрқайсысына -5%
    confidence -= anticheat["fast_answers"] * 5

    # Төмен variance: -20%
    if anticheat["variance_score"] < 0.3:
        confidence -= 20

    # Паттерн: -25%
    if anticheat["pattern_detected"]:
        confidence -= 25

    return max(0, min(100, confidence))


def format_confidence(confidence: int) -> str:
    """Сенімділікті HTML форматта көрсету."""
    filled = confidence // 20
    empty = 5 - filled

    if confidence >= 80:
        emoji = "🟢"
        icon = "🔒"
    elif confidence >= 50:
        emoji = "🟡"
        icon = "⚠️"
    else:
        emoji = "🔴"
        icon = "⚠️"

    bar = emoji * filled + "⚪" * empty
    text = f"{icon} <b>Нәтиже сенімділігі:</b> {confidence}% {bar}"

    if confidence < 50:
        text += "\n💡 <i>Сұрақтарға мұқият жауап беріп, қайта тест тапсырыңыз.</i>"

    return text


def _longest_consecutive_run(values: list) -> int:
    """Қатарынан қанша бірдей мән бар."""
    if not values:
        return 0
    max_run = 1
    current_run = 1
    for i in range(1, len(values)):
        if values[i] == values[i - 1]:
            current_run += 1
            max_run = max(max_run, current_run)
        else:
            current_run = 1
    return max_run


def _detect_cycling_pattern(values: list, min_period: int = 2, max_period: int = 5) -> bool:
    """Қысқа периодпен қайталанатын паттернді анықтау."""
    for period in range(min_period, max_period + 1):
        if len(values) < period * 2:
            continue
        pattern = values[:period]
        matches = sum(
            1 for i in range(period, len(values))
            if values[i] == pattern[i % period]
        )
        match_ratio = matches / (len(values) - period)
        if match_ratio >= 0.8:
            return True
    return False

File: frontend/src/utils/antiCheat.js
export function calculateConfidence(answerList, timingList) {
  let confidence = 100;
  const indices = answerList.map(a => a.option_index);

  // Straight-line
  let maxRun = 1, run = 1;
  for (let i = 1; i < indices.length; i++) {
    if (indices[i] === indices[i - 1]) { run++; maxRun = Math.max(maxRun, run); }
    else run = 1;
  }
  if (maxRun >= 4) confidence -= (maxRun - 3) * 15;

  // Speed
  const fastCount = timingList.filter(t => t < 2000).length;
  confidence -= fastCount * 5;

  // Pattern cycling
  for (let period = 2; period <= 5; period++) {
    if (indices.length < period * 2) continue;
    const pattern = indices.slice(0, period);
    let matches = 0;
    for (let i = period; i < indices.length; i++) {
      if (indices[i] === pattern[i % period]) matches++;
    }
    if (matches / (indices.length - period) >= 0.8) { confidence -= 25; break; }
  }

  // Low variance
  const unique = new Set(indices).size;
  if (unique <= 2 && indices.length >= 10) confidence -= 20;

  return Math.max(0, Math.min(100, confidence));
}

export function checkLiveAntiCheat(answers) {
  if (answers.length < 4) return null;
  const last4 = answers.slice(-4).map(a => a.option_index);
  if (last4.every(v => v === last4[0])) return 'Бірдей жауаптар байқалды!';
  const isAlt = last4[0] === last4[2] && last4[1] === last4[3] && last4[0] !== last4[1];
  if (isAlt) return 'Паттерн байқалды!';
  return null;
}
